File: client_example.py
"""
Example client for calling the SAM2 segmentation endpoint
"""
import requests
import base64
from pathlib import Path
import json


def segment_image(image_path: str, endpoint_url: str, point_coords=None, point_labels=None, mode="point"):
    """
    Send an image to the Modal endpoint for segmentation
    
    Args:
        image_path: Path to the image file
        endpoint_url: Your Modal endpoint URL
        point_coords: List of [x, y] coordinates (optional)
        point_labels: List of labels (1=foreground, 0=background) (optional)
        mode: "point" for point-based or "auto" for automatic segmentation
    """
    # Read and encode image
    with open(image_path, "rb") as f:
        image_bytes = f.read()
    
    image_base64 = base64.b64encode(image_bytes).decode("utf-8")
    
    # Prepare request payload
    payload = {
        "image": image_base64,
        "mode": mode,
    }
    
    if point_coords:
        payload["point_coords"] = point_coords
    if point_labels:
        payload["point_labels"] = point_labels
    
    # Send request
    response = requests.post(endpoint_url, json=payload)
    
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Error: {response.status_code} - {response.text}")


# Example usage
if __name__ == "__main__":
    # Replace with your Modal endpoint URL after deployment
    ENDPOINT_URL = "https://your-username--sam2-segmentation-segment-endpoint.modal.run"
    
    # Example 1: Point-based segmentation
    result = segment_image(
        image_path="test_image.jpg",
        endpoint_url=ENDPOINT_URL,
        point_coords=[[100, 100]],  # Click at position (100, 100)
        point_labels=[1],  # Foreground point
        mode="point"
    )
    print("Point-based segmentation result:")
    print(f"Score: {result['score']}")
    print(f"Image shape: {result['image_shape']}")
    
    # Example 2: Automatic segmentation (segment everything)
    result_auto = segment_image(
        image_path="test_image.jpg",
        endpoint_url=ENDPOINT_URL,
        mode="auto"
    )
    print(f"\nAutomatic segmentation found {result_auto['count']} objects")

File: client_example.js
/**
 * JavaScript client for SAM2 segmentation endpoint
 * Works in both Node.js and browser environments
 */

/**
 * Convert File or Blob to base64
 */
async function fileToBase64(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      const base64 = reader.result.split(',')[1]; // Remove data:image/...;base64, prefix
      resolve(base64);
    };
    reader.onerror = reject;
    reader.readAsDataURL(file);
  });
}

/**
 * Segment an image using the Modal endpoint
 */
async function segmentImage(imageFile, options = {}) {
  const {
    endpointUrl,
    pointCoords = null,
    pointLabels = null,
    mode = 'point'
  } = options;

  // Convert image to base64
  const imageBase64 = await fileToBase64(imageFile);

  // Prepare payload
  const payload = {
    image: imageBase64,
    mode: mode
  };

  if (pointCoords) {
    payload.point_coords = pointCoords;
  }
  if (pointLabels) {
    payload.point_labels = pointLabels;
  }

  // Send request
  const response = await fetch(endpointUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  return await response.json();
}

/**
 * Example usage in a React component
 */
/*
import React, { useState } from 'react';

function ImageSegmentation() {
  const [result, setResult] = useState(null);
  const [loading, setLoading] = useState(false);
  const ENDPOINT_URL = 'https://your-username--sam2-segmentation-segment-endpoint.modal.run';

  const handleImageUpload = async (event) => {
    const file = event.target.files[0];
    if (!file) return;

    setLoading(true);
    try {
      // Point-based segmentation
      const result = await segmentImage(file, {
        endpointUrl: ENDPOINT_URL,
        pointCoords: [[100, 100]], // Click coordinates
        pointLabels: [1], // Foreground
        mode: 'point'
      });
      
      setResult(result);
      console.log('Segmentation score:', result.score);
    } catch (error) {
      console.error('Segmentation failed:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleAutoSegment = async (event) => {
    const file = event.target.files[0];
    if (!file) return;

    setLoading(true);
    try {
      // Automatic segmentation
      const result = await segmentImage(file, {
        endpointUrl: ENDPOINT_URL,
        mode: 'auto'
      });
      
      setResult(result);
      console.log('Found', result.count, 'objects');
    } catch (error) {
      console.error('Segmentation failed:', error);
    } finally {
      setLoading(false);
    }
  };

  return (
    <div>
      <h2>Image Segmentation</h2>
      <input 
        type="file" 
        accept="image/*" 
        onChange={handleImageUpload}
        disabled={loading}
      />
      <button onClick={handleAutoSegment} disabled={loading}>
        Auto Segment
      </button>
      {loading && <p>Processing...</p>}
      {result && (
        <div>
          <p>Segmentation complete!</p>
          <pre>{JSON.stringify(result, null, 2)}</pre>
        </div>
      )}
    </div>
  );
}

export default ImageSegmentation;
*/

// For Node.js usage
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { segmentImage, fileToBase64 };
}
